### ISSUE TITLE
ESP32-S3: Crash into the HardFault_Handler

### ISSUE THREAD

**Issue by DJDevon3** [adafruit/circuitpython#6791]
### CircuitPython version

```python
Adafruit CircuitPython 7.3.2 on 2022-07-20; FeatherS3 with ESP32S3
Board ID:unexpectedmaker_feathers3

and 

Adafruit CircuitPython 8.0.0-beta.0 on 2022-08-18; FeatherS3 with ESP32S3
Board ID:unexpectedmaker_feathers3
```

### Code/REPL

```python
# SPDX-FileCopyrightText: 2022 DJDevon3 for Adafruit Industries
# SPDX-License-Identifier: MIT
"""DJDevon3 UM ESP32 Feather S3 Online Weatherstation"""
import gc
import time
import board
import feathers3
import displayio
import terminalio
from adafruit_display_text import label
from adafruit_bitmap_font import bitmap_font
from adafruit_bitmap_font import bitmap_font
import adafruit_imageload
import ssl
import wifi
import socketpool
import adafruit_requests
from adafruit_bme280 import basic as adafruit_bme280
from adafruit_hx8357 import HX8357

# 3.5" TFT Featherwing is 480x320
displayio.release_displays()
DISPLAY_WIDTH = 480
DISPLAY_HEIGHT = 320
sleep_time = 3600  # Time between weather updates

# Initialize TFT Display
spi = board.SPI()
tft_cs = board.IO1
tft_dc = board.IO3
display_bus = displayio.FourWire(spi,
                                 command=tft_dc,
                                 chip_select=tft_cs,
                                 baudrate=50000000,
                                 phase=0,
                                 polarity=0)
display = HX8357(display_bus, width=DISPLAY_WIDTH, height=DISPLAY_HEIGHT)

# Initialize BME280 Sensor
i2c = board.I2C()  # uses board.SCL and board.SDA
bme280 = adafruit_bme280.Adafruit_BME280_I2C(i2c)

Bat_S3 = feathers3.get_battery_voltage()
# print(Bat_S3)

try:
    from secrets import secrets
except ImportError:
    print("Secrets File Import Error")
    raise

# Friendly time instead of seconds for next weather update
if (sleep_time) < 60:
    sleep_time_conversion = "seconds"
    sleep_int = sleep_time
elif (sleep_time) >= 60 and (sleep_time) < 3600:
    sleep_int = sleep_time / 60
    sleep_time_conversion = "minutes"
elif (sleep_time) >= 3600:
    sleep_int = sleep_time / 60 / 60
    sleep_time_conversion = "hours"

# Quick Colors for Labels
text_black = 0x000000
text_blue = 0x0000FF
text_cyan = 0x00FFFF
text_gray = 0x8B8B8B
text_green = 0x00FF00
text_lightblue = 0x90C7FF
text_magenta = 0xFF00FF
text_orange = 0xFFA500
text_purple = 0x800080
text_red = 0xFF0000
text_white = 0xFFFFFF
text_yellow = 0xFFFF00

# Fonts are optional
medium_font = bitmap_font.load_font("/fonts/Arial-16.bdf")
huge_font = bitmap_font.load_font("/fonts/GoodTimesRg-Regular-121.bdf")

# Fill OpenWeather 2.5 API with token data
# OpenWeather free account & token are required
timezone = secrets['timezone']
tz_offset_seconds = int(secrets['timezone_offset'])

# OpenWeather 2.5 Free API
# Map source currently unused because PNG
MAP_SOURCE = "https://tile.openweathermap.org/map"
MAP_SOURCE += "/precipitation_new"
MAP_SOURCE += "/9"
MAP_SOURCE += "/"+secrets['openweather_lat']
MAP_SOURCE += "/"+secrets['openweather_lon']
MAP_SOURCE += "/.png"
MAP_SOURCE += "&appid="+secrets['openweather_token']

DATA_SOURCE = "https://api.openweathermap.org/data/2.5/onecall?"
DATA_SOURCE += "lat="+secrets['openweather_lat']
DATA_SOURCE += "&lon="+secrets['openweather_lon']
DATA_SOURCE += "&timezone="+timezone
DATA_SOURCE += "&timezone_offset="+str(tz_offset_seconds)
DATA_SOURCE += "&exclude=hourly,daily"
DATA_SOURCE += "&appid="+secrets['openweather_token']
DATA_SOURCE += "&units=imperial"

def _format_datetime(datetime):
    return "{:02}/{:02}/{} {:02}:{:02}:{:02}".format(
        datetime.tm_mon,
        datetime.tm_mday,
        datetime.tm_year,
        datetime.tm_hour,
        datetime.tm_min,
        datetime.tm_sec,
    )

def _format_date(datetime):
    return "{:02}/{:02}/{:02}".format(
        datetime.tm_year,
        datetime.tm_mon,
        datetime.tm_mday,
    )

def _format_time(datetime):
    return "{:02}:{:02}".format(
        datetime.tm_hour,
        datetime.tm_min,
        # datetime.tm_sec,
    )

# Weatherstation environment display labels
date_label = label.Label(medium_font)
date_label.anchor_point = (0.0, 0.0)
date_label.anchored_position = (5, 5)
date_label.scale = 1
date_label.color = text_lightblue

time_label = label.Label(medium_font)
time_label.anchor_point = (0.0, 0.0)
time_label.anchored_position = (5, 25)
time_label.scale = 2
time_label.color = text_lightblue

temp_label = label.Label(medium_font)
temp_label.anchor_point = (1.0, 1.0)
temp_label.anchored_position = (475, 145)
temp_label.scale = 2
temp_label.color = text_white

temp_data_label = label.Label(huge_font)
temp_data_label.anchor_point = (0.5, 1.0)
temp_data_label.anchored_position = (DISPLAY_WIDTH / 2, 200)
temp_data_label.scale = 1
temp_data_label.color = text_orange

temp_data_shadow = label.Label(huge_font)
temp_data_shadow.anchor_point = (0.5, 1.0)
temp_data_shadow.anchored_position = (DISPLAY_WIDTH / 2 + 2, 200 + 2)
temp_data_shadow.scale = 1
temp_data_shadow.color = text_black

owm_temp_data_label = label.Label(medium_font)
owm_temp_data_label.anchor_point = (0.5, 1.0)
owm_temp_data_label.anchored_position = (DISPLAY_WIDTH / 2, 100)
owm_temp_data_label.scale = 2
owm_temp_data_label.color = text_lightblue

owm_temp_data_shadow = label.Label(medium_font)
owm_temp_data_shadow.anchor_point = (0.5, 1.0)
owm_temp_data_shadow.anchored_position = (DISPLAY_WIDTH / 2 + 2, 100 + 2)
owm_temp_data_shadow.scale = 2
owm_temp_data_shadow.color = text_black

humidity_label = label.Label(medium_font)
humidity_label.anchor_point = (0.0, 1.0)
humidity_label.anchored_position = (5, DISPLAY_HEIGHT - 23)
humidity_label.scale = 1
humidity_label.color = text_gray

humidity_data_label = label.Label(medium_font)
humidity_data_label.anchor_point = (0.0, 1.0)
humidity_data_label.anchored_position = (5, DISPLAY_HEIGHT)
humidity_data_label.scale = 1
humidity_data_label.color = text_orange

owm_humidity_data_label = label.Label(medium_font)
owm_humidity_data_label.anchor_point = (0.0, 1.0)
owm_humidity_data_label.anchored_position = (5, DISPLAY_HEIGHT - 55)
owm_humidity_data_label.scale = 1
owm_humidity_data_label.color = text_lightblue

barometric_label = label.Label(medium_font)
barometric_label.anchor_point = (1.0, 1.0)
barometric_label.anchored_position = (470, DISPLAY_HEIGHT - 27)
barometric_label.scale = 1
barometric_label.color = text_gray

barometric_data_label = label.Label(medium_font)
barometric_data_label.anchor_point = (1.0, 1.0)
barometric_data_label.anchored_position = (470, DISPLAY_HEIGHT)
barometric_data_label.scale = 1
barometric_data_label.color = text_orange

owm_barometric_data_label = label.Label(medium_font)
owm_barometric_data_label.anchor_point = (1.0, 1.0)
owm_barometric_data_label.anchored_position = (470, DISPLAY_HEIGHT - 55)
owm_barometric_data_label.scale = 1
owm_barometric_data_label.color = text_lightblue

# Battery gauge display labels
vbat_label = label.Label(medium_font)
vbat_label.anchor_point = (1.0, 1.0)
vbat_label.anchored_position = (DISPLAY_WIDTH-15, 20)
vbat_label.scale = (1)

plugbmp_label = label.Label(terminalio.FONT)
plugbmp_label.anchor_point = (1.0, 1.0)
plugbmp_label.scale = (1)

greenbmp_label = label.Label(terminalio.FONT)
greenbmp_label.anchor_point = (1.0, 1.0)
greenbmp_label.scale = (1)

bluebmp_label = label.Label(terminalio.FONT)
bluebmp_label.anchor_point = (1.0, 1.0)
bluebmp_label.scale = (1)

yellowbmp_label = label.Label(terminalio.FONT)
yellowbmp_label.anchor_point = (1.0, 1.0)
yellowbmp_label.scale = (1)

orangebmp_label = label.Label(terminalio.FONT)
orangebmp_label.anchor_point = (1.0, 1.0)
orangebmp_label.scale = (1)

redbmp_label = label.Label(terminalio.FONT)
redbmp_label.anchor_point = (1.0, 1.0)
redbmp_label.scale = (1)

# Load Bitmap to tile grid first (Background layer)
DiskBMP = displayio.OnDiskBitmap("/images/Astral_Fruit_8bit.bmp")
tile_grid = displayio.TileGrid(
    DiskBMP,
    pixel_shader=DiskBMP.pixel_shader,
    width=1,
    height=1,
    tile_width=DISPLAY_WIDTH,
    tile_height=DISPLAY_HEIGHT)

# Load battery voltage icons (from 1 sprite sheet image)
sprite_sheet, palette = adafruit_imageload.load("/icons/vbat_spritesheet.bmp",
                                                bitmap=displayio.Bitmap,
                                                palette=displayio.Palette)
sprite = displayio.TileGrid(sprite_sheet, pixel_shader=palette,
                            width=1,
                            height=1,
                            tile_width=11,
                            tile_height=20)
sprite_group = displayio.Group(scale=1)
sprite_group.append(sprite)
sprite_group.x = 470
sprite_group.y = 0

# Create subgroups
text_group = displayio.Group()
text_group.append(tile_grid)
temp_group = displayio.Group()
main_group = displayio.Group()

# Add subgroups to main display group
main_group.append(text_group)
main_group.append(temp_group)
main_group.append(sprite_group)

# Label Display Group (foreground layer)
text_group.append(date_label)
text_group.append(time_label)
temp_group.append(temp_label)
temp_group.append(temp_data_shadow)
temp_group.append(temp_data_label)
temp_group.append(owm_temp_data_shadow)
temp_group.append(owm_temp_data_label)
text_group.append(humidity_label)
text_group.append(humidity_data_label)
text_group.append(owm_humidity_data_label)
text_group.append(barometric_label)
text_group.append(barometric_data_label)
text_group.append(owm_barometric_data_label)
text_group.append(vbat_label)
text_group.append(plugbmp_label)
text_group.append(greenbmp_label)
text_group.append(bluebmp_label)
text_group.append(yellowbmp_label)
text_group.append(orangebmp_label)
text_group.append(redbmp_label)
display.show(main_group)

# Connect to WiFi
print("\n===============================")
print("Connecting to WiFi...")
pool = socketpool.SocketPool(wifi.radio)
requests = adafruit_requests.Session(pool, ssl.create_default_context())
wifi.radio.connect(secrets["ssid"], secrets["password"])
print("Connected!\n")

vbat_label.text = "{:.2f}".format(Bat_S3)
# source_index = 0

while True:
    gc.collect()

    # Change battery voltage label color with charge level
    if vbat_label.text >= "4.23":
        vbat_label.color = text_white
        sprite[0] = 3
    elif vbat_label.text >= "4.10" and vbat_label.text <= "4.22":
        vbat_label.color = text_green
        sprite[0] = 1
    elif vbat_label.text >= "4.00" and vbat_label.text <= "4.09":
        vbat_label.color = text_lightblue
        sprite[0]
    elif vbat_label.text >= "3.90" and vbat_label.text <= "3.99":
        vbat_label.color = text_yellow
        sprite[0] = 5
    elif vbat_label.text >= "3.80" and vbat_label.text <= "3.89":
        vbat_label.color = text_orange
        sprite[0] = 2
    elif vbat_label.text <= "3.79":
        vbat_label.color = text_red
        sprite[0] = 4
    else:
        vbat_label.color = text_white

    # tile_grid[0] = source_index % 21
    # source_index += 1

    vbat_label.text

    temp_label.text = "°F"
    temperature = "{:.1f}".format(bme280.temperature * 1.8 + 32)
    temp_data_shadow.text = temperature
    temp_data_label.text = temperature
    humidity_label.text = "Humidity"
    humidity_data_label.text = "{:.1f} %".format(bme280.relative_humidity)
    barometric_label.text = "Pressure"
    barometric_data_label.text = "{:.1f}".format(bme280.pressure)

    try:
        print("Attempting to GET Weather!")
        # Uncomment line below to print API URL with all data filled in
        #print("Full API GET URL: ", DATA_SOURCE)
        print("\n===============================")
        response = requests.get(DATA_SOURCE).json()

        # uncomment the 2 lines below to see full json response
        # dump_object = json.dumps(response)
        # print("JSON Dump: ", dump_object)

        get_timestamp = int(response['current']['dt'] + tz_offset_seconds)
        current_unix_time = time.localtime(get_timestamp)
        current_struct_time = time.struct_time(current_unix_time)
        current_date = "{}".format(_format_date(current_struct_time))
        current_time = "{}".format(_format_time(current_struct_time))

        sunrise = int(response['current']['sunrise'] + tz_offset_seconds)
        sunrise_unix_time = time.localtime(sunrise)
        sunrise_struct_time = time.struct_time(sunrise_unix_time)
        sunrise_time = "{}".format(_format_time(sunrise_struct_time))

        sunset = int(response['current']['sunset'] + tz_offset_seconds)
        sunset_unix_time = time.localtime(sunset)
        sunset_struct_time = time.struct_time(sunset_unix_time)
        sunset_time = "{}".format(_format_time(sunset_struct_time))

        owm_temp = response['current']['temp']
        owm_pressure = response['current']['pressure']
        owm_humidity = response['current']['humidity']
        weather_type = response['current']['weather'][0]['main']

        print("Timestamp:", current_date + " " + current_time)
        print("Sunrise:", sunrise_time)
        print("Sunset:", sunset_time)
        print("Temp:", owm_temp)
        print("Pressure:", owm_pressure)
        print("Humidity:", owm_humidity)
        print("Weather Type:", weather_type)

        print("\nNext Update in %s %s" % (int(sleep_int), sleep_time_conversion))
        print("===============================")

        gc.collect()
        date_label.text = current_date
        time_label.text = current_time
        owm_temp_data_shadow.text = "{:.1f}".format(owm_temp)
        owm_temp_data_label.text = "{:.1f}".format(owm_temp)
        owm_humidity_data_label.text = "{:.1f} %".format(owm_humidity)
        owm_barometric_data_label.text = "{:.1f}".format(owm_pressure)

    except (ValueError, RuntimeError) as e:
        print("Failed to get data, retrying\n", e)
        wifi.reset()
        time.sleep(60)
        continue
    response = None
    time.sleep(sleep_time)

```

### Behavior

Auto-reload is off.
Running in safe mode! Not running saved code.

You are in safe mode because:
CircuitPython core code crashed hard. Whoops!
Crash into the HardFault_Handler.
Please file an issue with the contents of your CIRCUITPY drive at 
https://github.com/adafruit/circuitpython/issues

Press any key to enter the REPL. Use CTRL-D to reload.

Adafruit CircuitPython 7.3.2 on 2022-07-20; FeatherS3 with ESP32S3

### Description

Issue documented in discord dev channel. https://discord.com/channels/327254708534116352/327298996332658690/1010913898708336681

### Additional information

Issue is worse on 8.0 beta than 7.3 because it still crashes to safe mode but without any error describing it in repl like it shows in 7.3. but issue occurs in both UF2 versions. 

after a certain period of time when i hit save in Mu the USB drive disconnects, reconnects, i pull up repl immediately, and it shows that message. usually about 2-3 minutes. i have to physically hit reset on the board, get to save my code once or twice before 2-3 minutes, and the process repeats itself. It’s a safe mode crashing loop. 

Here's the full contents of my CIRCUITPY drive, including /lib and everything. https://github.com/DJDevon3/CircuitPython/tree/main/UM_ESP32-S3%20Online-Offline%20Weatherstation

**Comment by DJDevon3** [comment]
Also I’m running almost identical code on another 3.5” tft featherwing except on the bluefruit sense and esp32 airlift. That one works no problem.

**Comment by DJDevon3** [comment]
Lost power during a storm today. Restarted PC. The device is unrecognized in Windows. Throws an error when unplug/replug.

![USB Malfunction](https://user-images.githubusercontent.com/49322231/187040941-1773ab9e-ddce-45d7-aa23-93b3f101191e.PNG)

Intentionally didn't hit the reset button to collect info on this issue.
![IMG_0426](https://user-images.githubusercontent.com/49322231/187040965-a6d8698c-3282-49fb-bcc0-88c3743691a9.JPG)

Hit reset on the board.
![Drive_Problem](https://user-images.githubusercontent.com/49322231/187041017-b449ebde-9b06-4856-a06f-5d7a6fa8bb47.PNG)

Open Mu
![NoMu](https://user-images.githubusercontent.com/49322231/187041043-73dbc0ab-874f-4166-905e-a2c230a1057b.PNG)

Load code.py manually
![OpenManually](https://user-images.githubusercontent.com/49322231/187041082-230d7669-23c9-43ee-8e31-8c71c3b53062.PNG)

Pops up with another USB warning about the drive having a problem.

Open Serial Console and immediately greeted with this message.
![Serial](https://user-images.githubusercontent.com/49322231/187041118-bcc77418-46cf-41e5-b989-249ac717cc42.PNG)

Ctrl+D & Ctrl+C only reloads the safe mode message.  Cannot use Mu. 

Reset the board a 2nd time... works.
![After 2nd reset](https://user-images.githubusercontent.com/49322231/187041254-47d9d817-eab9-4371-a6f3-a3070736c22d.PNG)

If I attempt to save any code from here it will crash to safe mode again, this is the entire safe mode crash loop experience.

**Comment by DJDevon3** [comment]
Narrowed it down to the wifi or online portions of the code causing it. In it's offline state with no wifi code in it, it's works perfectly fine, can upload code with Mu and use REPL without issue.

**Comment by DJDevon3** [comment]
Ugh this is like hitting a moving target. Backed out of everything, returned to original code.py and everything works fine. This thing is time based, I have to wait like 10 minutes, put Mu in the background, then come back to it in order to replicate.

**Comment by DJDevon3** [comment]
Slimmed it down to this, saved, came back 10 minutes later, hit save again, works perfectly fine.

``` py
# SPDX-FileCopyrightText: 2022 DJDevon3 for Adafruit Industries
# SPDX-License-Identifier: MIT
"""DJDevon3 UM ESP32 Feather S3 Online Weatherstation"""

import board
import displayio
from adafruit_display_text import label
from adafruit_bitmap_font import bitmap_font
from adafruit_bme280 import basic as adafruit_bme280
from adafruit_hx8357 import HX8357

# 3.5" TFT Featherwing is 480x320
displayio.release_displays()
DISPLAY_WIDTH = 480
DISPLAY_HEIGHT = 320

# Initialize TFT Display
spi = board.SPI()
tft_cs = board.IO1
tft_dc = board.IO3
display_bus = displayio.FourWire(spi,
                                 command=tft_dc,
                                 chip_select=tft_cs,
                                 baudrate=50000000,
                                 phase=0,
                                 polarity=0)
display = HX8357(display_bus, width=DISPLAY_WIDTH, height=DISPLAY_HEIGHT)

# Initialize BME280 Sensor
i2c = board.I2C()  # uses board.SCL and board.SDA
bme280 = adafruit_bme280.Adafruit_BME280_I2C(i2c)

# Quick Colors for Labels
text_black = 0x000000
text_blue = 0x0000FF
text_cyan = 0x00FFFF
text_gray = 0x8B8B8B
text_green = 0x00FF00
text_lightblue = 0x90C7FF
text_magenta = 0xFF00FF
text_orange = 0xFFA500
text_purple = 0x800080
text_red = 0xFF0000
text_white = 0xFFFFFF
text_yellow = 0xFFFF00

# Fonts are optional
medium_font = bitmap_font.load_font("/fonts/Arial-16.bdf")
huge_font = bitmap_font.load_font("/fonts/GoodTimesRg-Regular-121.bdf")

temp_label = label.Label(medium_font)
temp_label.anchor_point = (1.0, 1.0)
temp_label.anchored_position = (475, 145)
temp_label.scale = 2
temp_label.color = text_white

temp_data_label = label.Label(huge_font)
temp_data_label.anchor_point = (0.5, 1.0)
temp_data_label.anchored_position = (DISPLAY_WIDTH / 2, 200)
temp_data_label.scale = 1
temp_data_label.color = text_orange

temp_data_shadow = label.Label(huge_font)
temp_data_shadow.anchor_point = (0.5, 1.0)
temp_data_shadow.anchored_position = (DISPLAY_WIDTH / 2 + 2, 200 + 2)
temp_data_shadow.scale = 1
temp_data_shadow.color = text_black

humidity_label = label.Label(medium_font)
humidity_label.anchor_point = (0.0, 1.0)
humidity_label.anchored_position = (5, DISPLAY_HEIGHT - 23)
humidity_label.scale = 1
humidity_label.color = text_gray

humidity_data_label = label.Label(medium_font)
humidity_data_label.anchor_point = (0.0, 1.0)
humidity_data_label.anchored_position = (5, DISPLAY_HEIGHT)
humidity_data_label.scale = 1
humidity_data_label.color = text_orange

barometric_label = label.Label(medium_font)
barometric_label.anchor_point = (1.0, 1.0)
barometric_label.anchored_position = (470, DISPLAY_HEIGHT - 27)
barometric_label.scale = 1
barometric_label.color = text_gray

barometric_data_label = label.Label(medium_font)
barometric_data_label.anchor_point = (1.0, 1.0)
barometric_data_label.anchored_position = (470, DISPLAY_HEIGHT)
barometric_data_label.scale = 1
barometric_data_label.color = text_orange

# Load Bitmap to tile grid first (Background layer)
DiskBMP = displayio.OnDiskBitmap("/images/Astral_Fruit_8bit.bmp")
tile_grid = displayio.TileGrid(
    DiskBMP,
    pixel_shader=DiskBMP.pixel_shader,
    width=1,
    height=1,
    tile_width=DISPLAY_WIDTH,
    tile_height=DISPLAY_HEIGHT)

# Create subgroups
text_group = displayio.Group()
text_group.append(tile_grid)
temp_group = displayio.Group()
main_group = displayio.Group()

# Add subgroups to main display group
main_group.append(text_group)
main_group.append(temp_group)

# Label Display Group (foreground layer)
temp_group.append(temp_label)
temp_group.append(temp_data_shadow)
temp_group.append(temp_data_label)
text_group.append(humidity_label)
text_group.append(humidity_data_label)
text_group.append(barometric_label)
text_group.append(barometric_data_label)
display.show(main_group)

while True:
    temp_label.text = "°F"
    temperature = "{:.1f}".format(bme280.temperature * 1.8 + 32)
    temp_data_shadow.text = temperature
    temp_data_label.text = temperature
    humidity_label.text = "Humidity"
    humidity_data_label.text = "{:.1f} %".format(bme280.relative_humidity)
    barometric_label.text = "Pressure"
    barometric_data_label.text = "{:.1f}".format(bme280.pressure)
```

**Comment by DJDevon3** [comment]
Ran some tests on Adafruit Feather ESP32-S2 with the same code. Issue is not present on the S2.

**Comment by DJDevon3** [comment]
Updated to nightly build hoping it would sort some stuff out after UM pushed a PR for LDO pin and some other bug hunt stuff going on recently. Still getting the same issue. Hasn't given me a hard fault crash yet. After a couple minutes after hard reset Mu doesn't seem to be able to save to the board. It's hanging during the process and waits to reload forever, indefinitely. Ctrl+C or Ctrl+D unresponsive.

Adafruit CircuitPython 8.0.0-beta.0-47-gae64f9fd7 on 2022-09-14; FeatherS3 with ESP32S3
Board ID:unexpectedmaker_feathers3

Code stopped by auto-reload. Reloading soon.

**Comment by DJDevon3** [comment]
Oops, was using 7.3.3 libraries. Added in 8.x libraries and it's working better. It refuses to reconnect to wifi if using Mu to soft-reload. It'll never connect. Hit reset and keep Mu closed and it will connect to wifi every time. Can only use Mu for serial monitoring after a hard reset. As soon as you hit save, auto reload, it'll never connect again.  

Haven't hit safe mode & hard fault with new 8.x 9/14 bundle libraries yet.

**Comment by dhalbert** [comment]
Based on your testing this afternoon, should we close this?

**Comment by DJDevon3** [comment]
I'm hesitant to close this just yet, only did about 5 mins worth of tests which isn't enough time for this bug especially knowing others are having very similar hard faults. It'll be great if all the hard faults end up being USB hub or controller issues. I'd like to spend at least a couple days with it just in case. Also have some more tests to do on the rp2040, m4 express, and S2 as well.

**Comment by DJDevon3** [comment]
Issue persists. Still crashing to safe mode and hard faulting.  Windows continually reports an error with the drive, there's some kind of issue with the way windows is handling this particular drive vs all of my other boards. It's also the only board that's crashing with hard faults.

**Comment by dhalbert** [comment]
Do you have another S3 board to test?

You might just try reformatting the filesystem too, with `storage.erase_filesystem()`.

**Comment by DJDevon3** [comment]
I have a lot of different boards I can test. RP2040, M0 Wifi, M4 Express, Bluefruit Sense, Pi Pico, ESP32-S2, QT Py S3.

**Comment by DJDevon3** [comment]
I do actually have another S3 but it's the same exact UM Feather S3. Bought 2.  Oh I've run it through the nuking process twice to no avail.

**Comment by DJDevon3** [comment]
Upon plugging in the 2nd UM FeatherS3 Windows immediately popped up the same message about there being a problem with the drive. Windows does not like this board. No errors in device manager.

Running the default code that comes with the UM FeatherS3 code.py
```py
import time, gc, os
import neopixel
import board, digitalio
import feathers3

# Create a NeoPixel instance
# Brightness of 0.3 is ample for the 1515 sized LED
pixel = neopixel.NeoPixel(board.NEOPIXEL, 1, brightness=0.3, auto_write=True, pixel_order=neopixel.RGB)

# Say hello
print("\nHello from FeatherS3!")
print("------------------\n")

# Show available memory
print("Memory Info - gc.mem_free()")
print("---------------------------")
print("{} Bytes\n".format(gc.mem_free()))

flash = os.statvfs('/')
flash_size = flash[0] * flash[2]
flash_free = flash[0] * flash[3]
# Show flash size
print("Flash - os.statvfs('/')")
print("---------------------------")
print("Size: {} Bytes\nFree: {} Bytes\n".format(flash_size, flash_free))

print("Pixel Time!\n")

# Create a colour wheel index int
color_index = 0

# Turn on the power to the NeoPixel
feathers3.set_ldo2_power(True)

# Rainbow colours on the NeoPixel
while True:
    # Get the R,G,B values of the next colour
    r,g,b = feathers3.rgb_color_wheel( color_index )
    # Set the colour on the NeoPixel
    pixel[0] = ( r, g, b, 0.5)
    # Increase the wheel index
    color_index += 1

    # If the index == 255, loop it
    if color_index == 255:
        color_index = 0
        # Invert the internal LED state every half colour cycle
        feathers3.led_blink()
        
    # Sleep for 15ms so the colour cycle isn't too fast
    time.sleep(0.015)
    
```
While the first hard faulting feather is plugged in I can use the 2nd feather no problem. 2nd feather seems fine. It does take much longer than normal for it to soft-reload.  Like 5-10 seconds. It's slow to boot compared to most other boards I own.

Pasted over the same script from the first feather and it ran. Throwing some errors for missing libs because I didn't have it plugged into the tft featherwing. Plugged into the tft featherwing, rebooted and it immediately crashed into the hard fault handler. Removed it from the featherwing. Removed all display related stuff from the script. Reset, ran again, still crashing to hard fault handler. Saw that I didn't unplug the crashing feather, transferred images/icons/font/libs to 2nd feather, then unplugged crashing feather... 

Same scripts and everything duplicated to the 2nd UM FeatherS3 works fine. Can soft reload no problem. REPL is responsive nothing out of the ordinary. Wait 5 minutes, hit save in Mu again on the 2nd S3 and it crashes to the HardFault_Hander. There is a problem with UM FeatherS3's.

Will try harder to narrow down the problem.

**Comment by dhalbert** [comment]
You might check with @UnexpectedMaker at their site about this.

**Comment by DJDevon3** [comment]
This code does not crash. Unused imports compared to crashing script are:
import terminalio
import adafruit_imageload
import ssl
import adafruit_requests
from adafruit_display_text import label
from adafruit_bitmap_font import bitmap_font

pretty sure I can remove all the display related stuff and it'll be fine.  

```py
# SPDX-FileCopyrightText: 2022 DJDevon3 for Adafruit Industries
# SPDX-License-Identifier: MIT
"""DJDevon3 UM ESP32 Feather S3 Online Weatherstation"""
import gc
import supervisor
import time
import board
import feathers3
import displayio
import digitalio
import adafruit_sdcard
import storage
import wifi
import socketpool
import adafruit_requests
from adafruit_hx8357 import HX8357

# USB Power Sensing
usb_sense = supervisor.runtime.serial_connected

# 3.5" TFT Featherwing is 480x320
displayio.release_displays()
DISPLAY_WIDTH = 480
DISPLAY_HEIGHT = 320

# Initialize WiFi Pool (This should always be near the top of a script!)
# anecdata: you only want to do this once early in your code pool.
# Highlander voice: "There can be only one pool"
pool = socketpool.SocketPool(wifi.radio)

# Time between weather updates
# 900 = 15 mins, 1800 = 30 mins, 3600 = 1 hour
sleep_time = 900

# Initialize TFT Display
spi = board.SPI()
tft_cs = board.IO1
tft_dc = board.IO3
display_bus = displayio.FourWire(spi,
                                 command=tft_dc,
                                 chip_select=tft_cs)
display = HX8357(display_bus, width=DISPLAY_WIDTH, height=DISPLAY_HEIGHT)


# Initialize SDCard on TFT Featherwing
cs = digitalio.DigitalInOut(board.IO33)
sdcard = adafruit_sdcard.SDCard(spi, cs)
vfs = storage.VfsFat(sdcard)
virtual_root = "/sd"
storage.mount(vfs, virtual_root)

Bat_S3 = feathers3.get_battery_voltage()
# print(Bat_S3)

try:
    from secrets import secrets
except ImportError:
    print("Secrets File Import Error")
    raise

if sleep_time < 60:
    sleep_time_conversion = "seconds"
    sleep_int = sleep_time
elif 60 <= sleep_time < 3600:
    sleep_int = sleep_time / 60
    sleep_time_conversion = "minutes"
elif 3600 <= sleep_time < 86400:
    sleep_int = sleep_time / 60 / 60
    sleep_time_conversion = "hours"
else:
    sleep_int = sleep_time / 60 / 60 / 24
    sleep_time_conversion = "days"

def _format_datetime(datetime):
    return "{:02}/{:02}/{} {:02}:{:02}:{:02}".format(
        datetime.tm_mon,
        datetime.tm_mday,
        datetime.tm_year,
        datetime.tm_hour,
        datetime.tm_min,
        datetime.tm_sec,
    )


def _format_date(datetime):
    return "{:02}/{:02}/{:02}".format(
        datetime.tm_year,
        datetime.tm_mon,
        datetime.tm_mday,
    )


def _format_time(datetime):
    return "{:02}:{:02}".format(
        datetime.tm_hour,
        datetime.tm_min,
        # datetime.tm_sec,
    )

while True:
    gc.collect()

    # Changes battery voltage color depending on charge level
    print("USB Sense: ", usb_sense)
    if usb_sense:
        print("USB connected")

    if not usb_sense:
        print("USB not connected")

    time.sleep(sleep_time)

```

**Comment by DJDevon3** [comment]
Tried narrowing it down more by removing all display related stuff.
Unused imports compared to original are:
import board
import displayio
import digitalio
import terminalio
import storage

```py
# SPDX-FileCopyrightText: 2022 DJDevon3 for Adafruit Industries
# SPDX-License-Identifier: MIT
"""DJDevon3 UM ESP32 Feather S3 Online Weatherstation"""
import gc
import supervisor
import time
import feathers3
import ssl
import wifi
import socketpool
import adafruit_requests


# USB Power Sensing
usb_sense = supervisor.runtime.serial_connected

# Initialize WiFi Pool (This should always be near the top of a script!)
# anecdata: you only want to do this once early in your code pool.
# Highlander voice: "There can be only one pool"
pool = socketpool.SocketPool(wifi.radio)

Bat_S3 = feathers3.get_battery_voltage()
# print(Bat_S3)

try:
    from secrets import secrets
except ImportError:
    print("Secrets File Import Error")
    raise

# Fill OpenWeather 2.5 API with token data
# OpenWeather free account & token are required
timezone = secrets['timezone']
tz_offset_seconds = int(secrets['timezone_offset'])

# OpenWeather 2.5 Free API
MAP_SOURCE = "https://tile.openweathermap.org/map"
MAP_SOURCE += "/precipitation_new"
MAP_SOURCE += "/9"
MAP_SOURCE += "/" + secrets['openweather_lat']
MAP_SOURCE += "/" + secrets['openweather_lon']
MAP_SOURCE += "/.png"
MAP_SOURCE += "&appid=" + secrets['openweather_token']

DATA_SOURCE = "https://api.openweathermap.org/data/2.5/onecall?"
DATA_SOURCE += "lat=" + secrets['openweather_lat']
DATA_SOURCE += "&lon=" + secrets['openweather_lon']
DATA_SOURCE += "&timezone=" + timezone
DATA_SOURCE += "&timezone_offset=" + str(tz_offset_seconds)
DATA_SOURCE += "&exclude=hourly,daily"
DATA_SOURCE += "&appid=" + secrets['openweather_token']
DATA_SOURCE += "&units=imperial"


def _format_datetime(datetime):
    return "{:02}/{:02}/{} {:02}:{:02}:{:02}".format(
        datetime.tm_mon,
        datetime.tm_mday,
        datetime.tm_year,
        datetime.tm_hour,
        datetime.tm_min,
        datetime.tm_sec,
    )


def _format_date(datetime):
    return "{:02}/{:02}/{:02}".format(
        datetime.tm_year,
        datetime.tm_mon,
        datetime.tm_mday,
    )


def _format_time(datetime):
    return "{:02}:{:02}".format(
        datetime.tm_hour,
        datetime.tm_min,
        # datetime.tm_sec,
    )

# Connect to Wi-Fi
print("\n===============================")
print("Connecting to WiFi...")
requests = adafruit_requests.Session(pool, ssl.create_default_context())
while not wifi.radio.ipv4_address:
    try:
        wifi.radio.enabled = False
        wifi.radio.enabled = True
        wifi.radio.connect(secrets['ssid'], secrets['password'])
    except ConnectionError as e:
        print("Connection Error:", e)
        print("Retrying in 10 seconds")
    time.sleep(10)
    gc.collect()
print("Connected!\n")


while True:
    gc.collect()

    # Changes battery voltage color depending on charge level
    print("USB Sense: ", usb_sense)
    if usb_sense:
        print("USB connected")

    if not usb_sense:
        print("USB disconnected")

    try:
        print("Attempting to GET Weather!")
        # Uncomment line below to print API URL with all data filled in
        # print("Full API GET URL: ", DATA_SOURCE)
        print("\n===============================")
        response = requests.get(DATA_SOURCE).json()

        # uncomment the 2 lines below to see full json response
        # dump_object = json.dumps(response)
        # print("JSON Dump: ", dump_object)
        if int(response['current']['dt']) == "KeyError: example":
            print("Unable to retrive data due to key error")
            print("most likely OpenWeather Throttling for too many API calls per day")
        else:
            print("OpenWeather Success")

        get_timestamp = int(response['current']['dt'] + tz_offset_seconds)
        current_unix_time = time.localtime(get_timestamp)
        current_struct_time = time.struct_time(current_unix_time)
        current_date = "{}".format(_format_date(current_struct_time))
        current_time = "{}".format(_format_time(current_struct_time))

        sunrise = int(response['current']['sunrise'] + tz_offset_seconds)
        sunrise_unix_time = time.localtime(sunrise)
        sunrise_struct_time = time.struct_time(sunrise_unix_time)
        sunrise_time = "{}".format(_format_time(sunrise_struct_time))

        sunset = int(response['current']['sunset'] + tz_offset_seconds)
        sunset_unix_time = time.localtime(sunset)
        sunset_struct_time = time.struct_time(sunset_unix_time)
        sunset_time = "{}".format(_format_time(sunset_struct_time))

        owm_temp = response['current']['temp']
        owm_pressure = response['current']['pressure']
        owm_humidity = response['current']['humidity']
        weather_type = response['current']['weather'][0]['main']

        print("Timestamp:", current_date + " " + current_time)
        print("Sunrise:", sunrise_time)
        print("Sunset:", sunset_time)
        print("Temp:", owm_temp)
        print("Pressure:", owm_pressure)
        print("Humidity:", owm_humidity)
        print("Weather Type:", weather_type)

        print("\nNext Update in %s %s" % (int(sleep_int), sleep_time_conversion))
        print("===============================")

        gc.collect()

    except (ValueError, RuntimeError) as e:
        print("Failed to get data, retrying\n", e)
        time.sleep(60)
        continue
    response = None

    time.sleep(sleep_time)
```
This produces the most interesting result yet. An error I've never seen before. Refuses to connect to wifi, maybe I missed something?

```py
===============================
Connecting to WiFi...
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
Retrying in 10 seconds```

The error handler is catching a 205... what's a 205?

**Comment by DJDevon3** [comment]
Narrowed down the 205 as much as possible
```py
# SPDX-FileCopyrightText: 2022 DJDevon3 for Adafruit Industries
# SPDX-License-Identifier: MIT
# UM ESP32 FeatherS3

import wifi
import socketpool

# Initialize WiFi Pool 
pool = socketpool.SocketPool(wifi.radio)

try:
    from secrets import secrets
    timezone = secrets['timezone']
    tz_offset_seconds = int(secrets['timezone_offset'])
except ImportError:
    print("Secrets File Import Error")
    raise

# Connect to Wi-Fi
print("\n===============================")
print("Connecting to WiFi...")
while not wifi.radio.ipv4_address:
    try:
        wifi.radio.connect(secrets['ssid'], secrets['password'])
    except ConnectionError as e:
        print("Connection Error:", e)
        print("Retrying in 10 seconds")
        
# With the 205 Error it never reaches below this point
print("Connected!\n")
```
Oh duh because it's being caught by the wifi error handler the 205 error is internally wifi library related. Well at least I was able to narrow that one down.

```py
===============================
Connecting to WiFi...
Connection Error: Unknown failure 205
Retrying in 10 seconds
Connection Error: Unknown failure 205
```

**Comment by UnexpectedMaker** [comment]
I'm kinda confused by this issue - is this about hard faults or wifi issues?
Re the hard faults - cant see that having anything to do with my boards. Sounds like a combo of Mu and Windows.
Does it hard fault when you don't use Mu? shut it down, use Notepad?

**Comment by UnexpectedMaker** [comment]
Re the Wifi - your last code example - does that work on the QTPy S3?
Can you take a photo of your board and wiring and everything for me?

**Comment by anecdata** [comment]
`205` is a quite unhelpful error code that arises from time to time:
`CONNECTION_FAIL`  "Espressif-specific Wi-Fi reason code: the connection to the AP has failed."
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html

**Comment by DJDevon3** [comment]
Yes the small sketch works on the Qt Py S3.

Adafruit CircuitPython 8.0.0-alpha.1 on 2022-06-09; Adafruit QT Py ESP32-S3 no psram with ESP32S3
Board ID:adafruit_qtpy_esp32s3_nopsram

```py
# secrets.py
secrets = {
    "ssid": "Your SSID",
    "password": "Your Wifi Password",
    "timezone": "America/New_York",  # Check http://worldtimeapi.org/timezones
    "aio_username": "AdafruitIO_Username",
    "aio_key": "AdafruitIO_Key",
    "timezone_offset": "-14400",  # New York timezone. offset in seconds plus or minus
    "openweather_token" : "long_token_here",
    "openweather_lat" : "27.44", #default lat/lon is adafruit industries
    "openweather_lon" : "-80.32",
}
```

```py
# code.py
import wifi
import socketpool

# Initialize WiFi Pool 
pool = socketpool.SocketPool(wifi.radio)

try:
    from secrets import secrets
    timezone = secrets['timezone']
    tz_offset_seconds = int(secrets['timezone_offset'])
except ImportError:
    print("Secrets File Import Error")
    raise

# Connect to Wi-Fi
print("\n===============================")
print("Connecting to WiFi...")
while not wifi.radio.ipv4_address:
    try:
        wifi.radio.connect(secrets['ssid'], secrets['password'])
    except ConnectionError as e:
        print("Connection Error:", e)
        print("Retrying in 10 seconds")
        
# With the 205 Error it never reaches below this point
print("Connected!\n")
```
Connected first attempted, consecutive reloads also connected.
```py
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:

===============================
Connecting to WiFi...
Connected!


Code done running.
```
Pictures of boards and setup soon to follow.

**Comment by DJDevon3** [comment]
![IMG_0483](https://user-images.githubusercontent.com/49322231/193440002-9013c784-14df-45f9-bc5e-d96371e430ad.jpg)
![IMG_0485](https://user-images.githubusercontent.com/49322231/193440017-e3463532-175b-49ac-83d1-3d1490d10bb7.jpg)
![IMG_0486](https://user-images.githubusercontent.com/49322231/193440004-eb5ec987-66ad-456a-8e68-b0333d269a4a.jpg)
![IMG_0487](https://user-images.githubusercontent.com/49322231/193440006-0a47e0b7-d35e-4dc9-b13c-085d1d903264.jpg)
![IMG_0488](https://user-images.githubusercontent.com/49322231/193440007-1dfac79a-f032-4986-a785-58a27b72c959.JPG)
![IMG_0489](https://user-images.githubusercontent.com/49322231/193440010-00bc715f-a1c5-4dae-a3d4-2a24e4b62255.JPG)
![IMG_0490](https://user-images.githubusercontent.com/49322231/193440012-f46c288b-8c82-4900-b29d-43fa20cec101.JPG)
![IMG_0491](https://user-images.githubusercontent.com/49322231/193440015-f2ee482d-ceb6-4198-b7a0-fb8d3d149ca0.JPG)

**Comment by UnexpectedMaker** [comment]
Interesting that worked for months for you - I know not your choice, but you have the hi speed signal wires of the screen sitting at the antenna - that's going to cause grief.  Plus it's inside a case, reducing the antenna gain even more, and you have an I2C device hanging off the back next to the antenna - it's a worst case setup there - I'm surprised that ever worked for you at all ;)

That said, if it was working - and now it's not...   I'm still perplexed as to what the reported problem is about.

Wifi was working fine and now isn't?  
Never had any hard faults but now there is?

If the HW hasn't changed though all of that...  and there have only been SW and firmware level changes - I'm not sure where I'm supposed to giving input on this?

**Comment by UnexpectedMaker** [comment]
It looks like your AP is just failing to assign the FS3 an IP address. 
Have you tried rebooting the router?
Have you tried assigning a static IP in the DHCP list for the FS3?

**Comment by DJDevon3** [comment]
Oh I said it was in that configuration since the day I got it. Didn't say it ever worked like the rest of my boards. I used to have an Adafruit S2 in there and that worked. Put your featherS3 in there and it's been non-stop reaching in there to hit the reset button. Can plug it in, run it once, and that works if I don't ever touch it again. That's how I have a pretty picture of it actually working. Try to work with the code in Mu or touch the USB drive to open a file on it and it crashes to the hard fault handler or wifi only works after hard reset (two different problems) at least in my main weatherstation sketch. It's like reaching into a bag of angry cats trying to get it to work right, only works right once after hard reset then crash into hard fault handler if I do anything else. 

It's NOT doing that with the tiny test sketch I just came up with in here, but it's also not connecting to wifi now which is the correlation.  I'm sure if I can get it to connect to wifi it'll crash into the hardfault_handler again. So yeah that's the main issue.

**Comment by UnexpectedMaker** [comment]
You wrote "First FeatherS3. Had it in this configuration for months. I love my little weatherstation. Just went through a hurricane with it btw."

Now you say it never worked properly?

I'm so confused and really have no idea where to go to from here.

**Comment by DJDevon3** [comment]
It works ONCE. If I touch the code or any file in USB it hard faults forcing a board reset.  So it works... and it doesn't.

**Comment by DJDevon3** [comment]
Now with the minimal sketch I've been able to connect reliably, waiting 5 minutes reloading, no errant behavior like I would normally get while it's connected to wifi in the main weatherstation sketch. Seems ok. Does this prove the culprit doesn't exist in minimal sketch? Now going to add adafruit_requests and try some basic GET requests to OpenWeather.

**Comment by UnexpectedMaker** [comment]
Remove Mu from all of this. Close it and keep it closed, and just use notepad.

Others have reported inconsistent issues using Mu on Windows. 1 person cant get the code.,py to reload when editing via Mu, and one cant always get the code.py to update on the device using Mu.

So please just use notepad and file explorer to test with.

Also, with WiFi, please ensure the channel you are using is 9 or greater for your testing.

**Comment by DJDevon3** [comment]
Set channel=11. Wifi only works after hard reset. If I change the file the wifi doesn't work, will search for wifi forever no ssid. 

Because I have it connected to a TFT I can usually see when it reloads, the screen will flash black, I'll see blinka in the top left corner, etc..  After sitting idle for a couple minutes, if I hit save in notepad++ nothing happens, no reload.

Hit reset, TFT reloads, searches and finds wifi, it will continue functioning normally in this state forever pulling info from OpenWeather every 15 minutes without fail.  

From here there are 2 situations that can happen:
1. Open code.py in notepad++ immediately after hard reset, hit save, no wifi.
2. Open code.py in notepad++, wait 5 minutes, hit save, no reload.

**Comment by UnexpectedMaker** [comment]
Can you please work on a board that is not in the case or connected to anything...   Too many variables, I want to focus on the wifi issues and the no reloading on a bare FS3.

**Comment by UnexpectedMaker** [comment]
I've had no reports of anything like this in any other dev platform other than CP - and only 3 instances of "code no reloading" as I mentioned above, and only in the last 2-3 weeks... and the FS3 has been out for > 8 months with no HW revisions.

So we need to track down what has changed from a CP point of view (IDF changes under the hood etc) and get some 100% reproducible steps under Windows and then see if they are reproducible on macOS and Linux.

**Comment by DJDevon3** [comment]
If I work on a board not connected to a TFT and use notepad++ there's no way to know if it's connecting to wifi, successful GET request, or crashing. Without REPL I'm working blind so I don't see what that's going to accomplish. I can use VScode with serial monitor but I have to reload serial monitor with every reset doing ctrl+d which reloads the code which in turn refuses to connect to wifi because it's not monitoring the first run after a hard reset.

**Comment by UnexpectedMaker** [comment]
So many options….
- use the rgb led to show connection status
- Use the blue led on IO13 to show status
- Use ‘tio’ to monitor serial output as it does auto reconnect
- Use mqtt or adafruitio and publish results online

**Comment by DJDevon3** [comment]
running this code now, let's call it minimal board only.
```py
import wifi
import socketpool
import board
from digitalio import DigitalInOut, Direction, Pull

# Initialize WiFi Pool
pool = socketpool.SocketPool(wifi.radio)

# Init Blink LED
led13 = DigitalInOut(board.LED)
led13.direction = Direction.OUTPUT

try:
    from secrets import secrets
except ImportError:
    print("Secrets File Import Error")
    raise

# Connect to Wi-Fi
print("\n===============================")
print("Connecting to WiFi...")
while not wifi.radio.ipv4_address:
    try:
        wifi.radio.connect(secrets['ssid'], secrets['password'])
        led13.value = True
    except ConnectionError as e:
        print("Connection Error:", e)

# With the 205 Error it never reaches below this point
print("Connected!\n")
```
Can't use tio in windows. Last time I tried it, it doesn't work and I uninstalled WSL trying to figure this stuff out.

**Comment by DJDevon3** [comment]
Installed the new 8.0.0.beta1 released last night. Getting a bunch of different results now.
```py
Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
]0;🐍Wi-Fi: off | code.py | 8.0.0-beta.1\]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.1\
===============================
Connecting to WiFi...
Connection Error: Unknown failure 4
Connection Error: Unknown failure 204
Connection Error: Unknown failure 204
Connection Error: Unknown failure 4
Connection Error: Unknown failure 204
Connection Error: Unknown failure 2
Connection Error: Unknown failure 4
Connection Error: Unknown failure 15
Connection Error: Unknown failure 4
Connection Error: Unknown failure 2
Connection Error: Unknown failure 4
Connection Error: Unknown failure 4
Connection Error: Unknown failure 4
Connection Error: Unknown failure 2
Connection Error: Unknown failure 2
Connection Error: Unknown failure 4
```

**Comment by DJDevon3** [comment]
It's gotta be something wrong with my code. Just ran the adafruit QT Py S3 internet test code and it works once only after hard reset. https://learn.adafruit.com/adafruit-qt-py-esp32-s3/circuitpython-internet-test

```py
]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.1\Connected to SSID!
My IP address is 192.168.1.2
Ping google.com: 0.000000 ms
Fetching text from http://wifitest.adafruit.com/testwifi/index.html
----------------------------------------
This is a test of Adafruit WiFi!
If you can read this, its working :)
----------------------------------------
Fetching json from https://www.adafruit.com/api/quotes.php
----------------------------------------
[{'text': 'We are the music makers, And we are the dreamers of dreams', 'author': "Arthur O'Shaughnessy"}]
----------------------------------------

Fetching and parsing json from https://api.github.com/repos/adafruit/circuitpython
----------------------------------------
CircuitPython GitHub Stars 3187
----------------------------------------
done
]0;🐍192.168.1.2 | Done | 8.0.0-beta.1\
Code done running.
]0;�Wi-Fi: off | Done | 8.0.0-beta.1\Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.

Press any key to enter the REPL. Use CTRL-D to reload.
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
]0;🐍Wi-Fi: off | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | Done | 8.0.0-beta.1\
Code stopped by auto-reload. Reloading soon.
]0;�Wi-Fi: off | Done | 8.0.0-beta.1\soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
]0;🐍Wi-Fi: off | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\ESP32-S2 WebClient Test
My MAC addr: ['0xf4', '0x12', '0xfa', '0x42', '0x4', '0x8']
Available WiFi networks:
]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.1\Connecting to Extra 2.4
Traceback (most recent call last):
  File "code.py", line 34, in <module>
]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.1\ConnectionError: No network with that ssid
]0;🐍Wi-Fi: No IP | 34@code.py ConnectionError | 8.0.0-beta.1\
Code done running.
]0;�Wi-Fi: off | Done | 8.0.0-beta.1\
Press any key to enter the REPL. Use CTRL-D to reload.
```
Connects once after hard reset, hit save, auto-reloads and fails to connect.

**Comment by UnexpectedMaker** [comment]
Sounds like CP is not handling the cleanup of wifi (or lower level) nicely on a soft reboot... could be related to this...
https://github.com/micropython/micropython/issues/8994

Might also be something else at the firmware level.

Have you tried rolling back to 7.3.0 or 7.2.x (cant remember what version initial S3 support arrived) and seeing if the same thing happens?

**Comment by UnexpectedMaker** [comment]
There is no wifi code cleanup at the end of that internet get code - I know the hole "should not need to" - but that assumes a device continuing to run... not a soft reset or even hard reset is reseting the PHY state properly.

What happens if you add `wifi.wifi_radio_set_enabled(False)` at the end of that code?

**Comment by Neradoc** [comment]
The autoreload issue is likely #6866 (which I tracked dosn to an ESP-IDF update - see my comment in the thread).

**Comment by UnexpectedMaker** [comment]
> The autoreload issue is likely #6866 (which I tracked down to an ESP-IDF update - see my comment in the thread).

Yeah, these issues have all become intertwined and hard to track/ cross reference - well at least for me ;)

**Comment by anecdata** [comment]
Some of these errors, namely the ConnectionError unknown errors 2, 4, 15, and 204 seem to me to indicate partial connectivity issues with the AP. What is the RSSI for the AP (same BSSID if multiple APs share an SSID) shown from the device point of view (either `wifi.radio.ap_info.rssi` if connected, or via a wifi scan)? What is the RSSI for the device that's hopefully shown in the GUI of the AP / router?

@DJDevon3  This discussion has gotten quite complex. It may help to summarize at this point your conclusions about which issue(s) are still occurring, whether each is correlated to particular CP version(s), and whether each is correlated to a particular port, board type, or specific board? Are any of the behaviors reproducible with a stripped-down hardware config and simple code... so that others could try it out? With CP 8, is web workflow enabled, with a `/env` file containing wifi credentials? If not, you may want to try some variations with web workflow enabled (it improves wifi scans and some other things, for as-yet mysterious reasons).

@UnexpectedMaker What is the significance of enforcing channel >= 9? I assume we have determined this is not antenna-related?

**Comment by UnexpectedMaker** [comment]
@anecdata The ESP32-S3 has a silicon bug when USB is being used with the radio on - there is a 2-3dB loss on lower channels... I Think 6 and below, so I just recommend folks go to 9 or higher if they can.

**Comment by DJDevon3** [comment]
https://github.com/espressif/arduino-esp32/issues/2501 found some evidence this is an esp32 interference & module init problem. I have about 3-4 esp32 running at any given time next to each other.  The handshake issue has existed for a while even affecting some magtags which I also have nearby. Also possible my router is having an issue with them. Hard to prove but can't rule out.

**Comment by DJDevon3** [comment]
Running S2 internet test found here: https://learn.adafruit.com/adafruit-esp32-s2-feather/circuitpython-internet-test

```py
]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\Connected to UAC 2.4!
My IP address is 192.168.1.215
Ping google.com: 139.000000 ms
Fetching text from http://wifitest.adafruit.com/testwifi/index.html
----------------------------------------
This is a test of Adafruit WiFi!
If you can read this, its working :)
----------------------------------------
Fetching json from https://www.adafruit.com/api/quotes.php
----------------------------------------
[{'text': 'As we enjoy great advantages from the inventions of others, we should be glad of an opportunity to serve others by any invention of ours; and this we should do freely and generously.', 'author': 'Benjamin Franklin'}]
----------------------------------------

Fetching and parsing json from https://api.github.com/repos/adafruit/circuitpython
----------------------------------------
CircuitPython GitHub Stars 3190
----------------------------------------
done
]0;🐍192.168.1.215 | Done | 8.0.0-beta.0-72-ga7b10d41b\
Code done running.
]0;�Wi-Fi: off | Done | 8.0.0-beta.0-72-ga7b10d41b\Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.

Press any key to enter the REPL. Use CTRL-D to reload.
```
OK that's weird. By the time it ends it automatically switches off wifi, is that normal for the test?  

Hit save to auto-reload and it produces this:
```py
soft reboot

Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
code.py output:
]0;🐍Wi-Fi: off | code.py | 8.0.0-beta.0-72-ga7b10d41b\]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\

ESP32-S2 WebClient Test
My MAC addr: ['0x84', '0xf7', '0x3', '0xd6', '0x8b', '0x72']
Available WiFi networks:
]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\	Extra 2.4		RSSI: -50	Channel: 6
	UAC 2.4		RSSI: -58	Channel: 6
]0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;�Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\]
0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\Connecting to UAC 2.4
]0;🐍Wi-Fi: No IP | code.py | 8.0.0-beta.0-72-ga7b10d41b\Traceback (most recent call last):
  File "code.py", line 31, in <module>
ConnectionError: Unknown failure 2
]0;🐍Wi-Fi: No IP | 31@code.py ConnectionError | 8.0.0-beta.0-72-ga7b10d41b\
Code done running.
]0;�Wi-Fi: off | 31@code.py ConnectionError | 8.0.0-beta.0-72-ga7b10d41b\
Press any key to enter the REPL. Use CTRL-D to reload.
```
This snake formatting has got to go. I cleaned it up a little but REPL output is just a mess now.

**Comment by anecdata** [comment]
When code exits, wifi deinits and the connection is dropped.

**Comment by DJDevon3** [comment]
Found possible old bug correlation https://github.com/adafruit/circuitpython/issues/3712 Todbot reported exact symptoms I have. Maybe the bug came back?

**Comment by anecdata** [comment]
My APs don't have DHCP either. Depending which AP a device connects to, it's path is: device --> one of several APs --> [some paths via one or more of several managed switches] --> router.

**Comment by anecdata** [comment]
Until 6866 is resolved, I'd suggest running CP8 as follows:

- run web workflow if possible, even if not needed functionally (I know you said there are issues there)
- avoid wifi scanning if possible, particularly after a reload (reset seems to be OK)
- insert the following before calls to `wifi.radio.connect()`:
```py
wifi.radio.enabled = False
wifi.radio.enabled = True
```

**Comment by DJDevon3** [comment]
I’ll have to add some exceptions to my firewall which is a pain to get at and in disarray after moving into a new place. Intend on running Ethernet this winter when my attic isn’t 120F death trap. Might take me a while to get at the firewall stuff.  

I can’t run web workflow wired and especially not wirelessly because port 80 is taken. Melissa’s guide says it doesn’t like running on anything but port 80 right now. That’s a problem for me and most lan admins even if it’s only on a loopback. I’d have to reorganize a good portion of my lan just to test web workflow. That’s beside the point of not being keen on poking holes in my firewall for IOT devices to access my maker pc and the internet.  While I love the idea of web workflow you have to admit it’s going to change the landscape of iot vlans, reconfiguration is required.  I wish it was as simple as just throwing creds into a env file but it’s really not. 

### Summary of this issue to date:
1. Hard reset bug in most cases hides the hard fault bug on the UM FeatherS3. Can’t get it to trigger reliably after updating to latest beta 1 released a few days ago. Progress. 
2. I cannot definitively say the issue is gone because I cannot keep WiFi running in a sketch past the 1st hard reset. Hard reset bug is obscuring hard fault bug. 

Hard fault bug investigation has to be put on hold until the hard reset bug is fixed. Too many variables and getting at it behind the hard reset bug not worth the insanity it’s causing.

**Comment by anecdata** [comment]
I infer from the guide that it's just the online Code Editor that hasn't been tested on other ports. The local file browser and code editor should be fine.

But you don't ewven need to use any of those tools to get the connection and scanning benefit. Simply having web workflow enabled seems to improve scanning and connection behavior. Web workflow could be enabled on an alternate port, like 8080, in the `/.env` file, then just forget web workflow exists and do things over USB or standalone as usual.

**Comment by DJDevon3** [comment]
The Oct 11th, 2022 nightly build fixed the hard-reset bug on the UM FeatherS3 as well. Pat on the back to DanH & anecdata, and anyone else involved in tracking it down. 

If this hard-fault bug (different than hard-reset bug) does still exist it should get exposed pretty quick in the next couple days. Because I'm using the UM FeatherS3 as my main board in my main long term project if it does hard fault I'll notice it immediately on my weatherstation display. 

I'd like a couple days to put it through paces, should be plenty of time. If all goes well I will gleefully close.

**Comment by DJDevon3** [comment]
Woke up this morning to find my weatherstation stopped at 11:30pm last night. Look at the modified timestamp of code.py and it says 10:06pm. It automatically updates every 15 minutes GET data from openweathermaps.  Only had a chance to successfully run about 3 times before crashing. The HardFault bug as I suspected persisted through the fix of the hard-reset bug fix.

```py
]0;🐍Wi-Fi: off | Done | 8.0.0-beta.1-29-gfed884738\Auto-reload is off.
Running in safe mode! Not running saved code.

You are in safe mode because:
CircuitPython core code crashed hard. Whoops!
Crash into the HardFault_Handler.
Please file an issue with the contents of your CIRCUITPY drive at 
https://github.com/adafruit/circuitpython/issues

Press any key to enter the REPL. Use CTRL-D to reload.
]0;🐍Wi-Fi: off | Done | 8.0.0-beta.1-29-gfed884738\
```

**Comment by DJDevon3** [comment]
RedBeard in discord suggested a possible correlation between I2C and HardFaults in multiple bug reports across the board. Unplugged I2C BME280, removed code, and set weather update interval to every 3 minutes beginning at 2:02pm.  It's now 2:23pm and no hard faults yet. If it is a memory leak of some kind it might take a while.

**Comment by DJDevon3** [comment]
Ran for 2 hours without issue. Tried removing some code to setup an I2C only example and it HardFaulted on hitting save in Mu.  It's got something to do with USB & file save after a certain period of time.

**Comment by DJDevon3** [comment]
Ran an I2C test for about an hour, hit save, reloaded without issue.  Issue is not solely I2C related.
```py
I2C BME280 Temp:  89.8
Monotonic 3079.88
I2C BME280 Temp:  89.8
Monotonic 3081.51
I2C BME280 Temp:  89.8
Monotonic 3083.08
```
I created a chart in an effort to convey the behavior on the UM FeatherS3 because it's really hard to explain. There is some kind of inactivity timer involved.
![Untitled](https://user-images.githubusercontent.com/49322231/195456382-1ee032ec-a093-471a-8486-976603ba87b3.png)

I don't have an exact time measurement because when it crashes it ejects the USB drive and wipes REPL.  I'd have to try to log REPL it to an SD Card? Unsure if going to that extreme is necessary but I'll do it if anyone thinks it'll help.

**Comment by DJDevon3** [comment]
Ported some of this project from the UM Feather S3 + TFT featherwing to an Adafruit Feather S3 4mb/2mb + RGBmatrix panel + RGB matrix featherwing.  It's also crashing into the hard fault handler. Reliably.  Only experimented with it for a couple hours last night. Initial impression is the timing related bug is still present on 8.0.0 beta 4. Fault handler crashes to safe mode which requires a hard reset. It only crashes when wifi is added to the code, was trying pull API data from an online source (openweathermaps).  Using wifi and displayio in both scripts/scenarios on different hardware. Still looking into it. Just an update from a beta testing perspective. Another correlation is after hard reset it will run no problem and trying to make a change and save in Mu triggers the hard fault.

**Comment by dhalbert** [comment]
Can you summarize the situations where it crashes?
1. If you start from a hard reset and never edit any files, does it keep running indefinitely?
2. If you start from a hard reset, and edit a file, so that it restarts, does it crash reliably? Immediately or much later?

**Comment by dhalbert** [comment]
Let's close this and open a new issue that's more precise if this comes up again. It's really hard to follow what's going on here.

**Comment by DJDevon3** [comment]
I haven't revisited my weather station project because of that bug, it completely derailed my project. The bug was in 7.x and continued into 8.x beta. We're now past the 8.x stable release date.  It's been permanently off for months.  Got sidetracked with other projects.  I'll make it a point to load this one back up tonight and test with the current stable version of 8.0.5.  I am still hesitant to close this issue as it was a doozy of a bug.  Will check and report back.

**Comment by DJDevon3** [comment]
I switched to the Adafruit ESP32-S3 which was exhibiting the same symptoms at the time during 7.x and 8.0 beta. Gave up on it at that point. 

```py
Adafruit CircuitPython 8.0.5 on 2023-03-31; Adafruit Feather ESP32S3 4MB Flash 2MB PSRAM with ESP32S3
Board ID:adafruit_feather_esp32s3_4mbflash_2mbpsram
```
Been running like a dream for an hour. Everything is much much faster too. I don't want to swap the UM FeatherS3 back in because it's deep inside an enclosure and would take about 20 minutes just to swap the boards.  

Can't say for certain but definitely feels like the hard fault bug with the S3's has been resolved judging from how the Adafruit ESP32-S3 is functioning. I'll take another shot at running a short demo on the UM FeatherS3 later.

The weather station TFT is sitting on my desk right in front of my main monitor 24/7. If it crops up again I'll definitely know it and will resubmit. 

Closing

**Comment by DJDevon3** [comment]
I did eventually get back to testing the UM FeatherS3 running 24 hours without a single error let alone a hard fault. This issue is now 100% confirmed closed on 8.0.5 stable release.
